### ISSUE TITLE
New client causes current client to wait infinitely

### ISSUE THREAD

**Issue by aliyehia92** [me-no-dev/ESPAsyncWebServer#579]
Hello.

First of all let me clarify that I am working with an ESP32 hooked to an ethernet PHY (LAN8720).

The issue I'm facing is basically this: if a client that is connected through ethernet tries to enable WiFi , and while WiFi is being enabled another client attempts to connect through ethernet, both clients come to a deadlock and the ESP completely stops responding to any requests until I reset it.

basically, I wrote an event handler for WiFi to see the steps it goes through while attempting to connect:

```
void WiFiEvent(WiFiEvent_t event, system_event_info_t info) {
	switch (event) {
	case SYSTEM_EVENT_STA_START:
		Serial.println("started");
		break;
	case SYSTEM_EVENT_STA_STOP:
		Serial.println("stopped");
		break;
	case SYSTEM_EVENT_STA_DISCONNECTED:
		Serial.println("couldn't connect");
		Serial.println("reason: ");
		Serial.println(info.disconnected.reason);
		wifi_connected = false;
		break;
	case SYSTEM_EVENT_STA_CONNECTED:
		Serial.print("Connected");
		break;
	case SYSTEM_EVENT_STA_GOT_IP:
		Serial.println("Wifi IP address: ");
		Serial.println(WiFi.localIP());
		wifi_connected = true;
		break;
	default:
		Serial.print("Event not handled:");
		Serial.println(event);
		break;
	}
}
```

So there's a boolean that is set to true when an IP address is obtained and is set to false when WiFi disconnects.

Now on to my setup() code. I'm going to skip over the Serial and SPIFFS initializations because they're trivial and very unlikely to be causing any damage.

The first thing I'm doing is starting up ethernet:
`ETH.begin(PHY1, 17, 23, 18, ETH_PHY_LAN8720, ETH_CLOCK_GPIO0_IN);`

Then, I'm allowing the user to turn WiFi on or off:
```
server.on("/", [](AsyncWebServerRequest *request) {
		request->send(200,"text/plain","/on or /off ");
	});
```

if the user chooses to turn it on, I'm checking if it's off first:
```
server.on("/on", [](AsyncWebServerRequest *request) {
		if(!wifi_connected) {
			WiFi.begin();
                       //loop infinitely until connection is established
                        while(WiFi.status() != WL_CONNECTED);
			wifi_connected = true;
		}
		request->send(200,"text/plain","on");
	});
```

same thing with turning it off:
```
server.on("/off", [](AsyncWebServerRequest *request) {
			if(wifi_connected) {
				WiFi.disconnect();
				wifi_connected = false;
			}
			request->send(200,"text/plain","off");
		});
```

then of course, start server:
`server.begin()`

That's basically it. and of course, my loop() is empty. Also bear in mind that the ssid and pass are already saved so I don't need to call begin() with parameters.
What's happening is that if I open my browser and visit "[ETH_IP]/on" and then open another tab and visit [ETH_IP], then there's a big chance that the WiFi will be stuck trying to connect.

What's confusing me is that the event handler would print "Connected" to the serial monitor but would never move on to obtaining IP phase. In fact, no other WiFi event is fired after SYSTEM_EVENT_STA_CONNECTED and the ESP can no longer respond to either client.

Why isn't it jumping from SYSTEM_EVENT_STA_CONNECTED to SYSTEM_EVENT_STA_GOT_IP which is what it normally does? 

I thought my code was pretty straight forward and I even previously tried sending simultaneous requests from two clients to the ESP but it did not crash and it served both clients with no issues. Am I missing something?

**Comment by aliyehia92** [comment]
Update:

I tried removing the code to connect to WiFi and replaced it instead with a simple loop and I'm still facing the same issue.

So what I did was I got rid of 
```
server.on("/on", [](AsyncWebServerRequest *request) {
		if(!wifi_connected) {
			WiFi.begin();
                       //loop infinitely until connection is established
                        while(WiFi.status() != WL_CONNECTED);
			wifi_connected = true;
		}
		request->send(200,"text/plain","on");
	});
```

and replaced it with
```
server.on("/on", [](AsyncWebServerRequest *request) {
		int i = 0;
		while(i<15){
			delay(500);
			Serial.print(i);
			i += 1;
		}
		request->send(200,"text/plain","ok");
	});
```

just to make sure that it's not enabling the WiFi that's causing problems, and I was right. 
What happens now is that i is printed from 0 to 14 and as soon as the loop ends everything is stuck like before. So I guess it's request->send() that's causing the issue.

**Comment by aliyehia92** [comment]
@atanisoft thanks a lot for your reply!  can you elaborate more on your answer? perhaps provide a code snippet? I'm not entirely sure how do that.

**Comment by aliyehia92** [comment]
my loop() has been empty so far and I don't want to write code inside it that would cause conflicts

**Comment by aliyehia92** [comment]
I see. thanks for the help!

**Comment by aliyehia92** [comment]
@atanisoft I just have one more question. So I followed your advice and did the below:

first I declared a global flag set to false:
`bool want_to_connect = false;`

and I changed my callback handler to the below:

```
server.on("/on", [](AsyncWebServerRequest *request) {
		request->send(200,"text/plain","/on or /off");
		want_to_connect = true;
		request->send(200,"text/plain","connected");


	});
```

and my loop() waits for the flag to become true to connect to WiFi:
```
void loop() {
	if (want_to_connect) {
		WiFi.begin();       
		while (WiFi.status() != WL_CONNECTED) {
			Serial.print(".");
			delay(500);
		}
		want_to_connect = false;
	}
}
```

which works fine. But what would be the best way to let the ESP hold until WiFi connects and then give feedback to the client? With the way I did it above the client would receive a confirmation before it even obtains an IP!

I suspect that declaring another flag and looping infinitely on it in the handler until it's set to true in the loop() is bad practice.
